src/ripple/consensus/LedgerTrie.h

```diff
@@ -0,0 +1,777 @@
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_APP_CONSENSUS_LEDGERS_TRIE_H_INCLUDED
+#define RIPPLE_APP_CONSENSUS_LEDGERS_TRIE_H_INCLUDED
+
+#include <ripple/json/json_value.h>
+#include <boost/optional.hpp>
+#include <algorithm>
+#include <memory>
+#include <vector>
+
+namespace ripple {
+
+namespace ledger_trie_detail {
+// Represents a span of ancestry of a ledger
+template <class Ledger>
+class Span
+{
+    using Seq = typename Ledger::Seq;
+    using ID = typename Ledger::ID;
+
+    // The span is the half-open interval [start,end) of ledger_
+    Seq start_{0};
+    Seq end_{1};
+    Ledger ledger_;
+
+public:
+    Span() : ledger_{typename Ledger::MakeGenesis{}}
+    {
+        // Require default ledger to be genesis seq
+        assert(ledger_.seq() == start_);
+    }
+
+    Span(Ledger ledger)
+        : start_{0}, end_{ledger.seq() + Seq{1}}, ledger_{std::move(ledger)}
+    {
+    }
+
+    Span(Span const& s) = default;
+    Span(Span&& s) = default;
+    Span&
+    operator=(Span const&) = default;
+    Span&
+    operator=(Span&&) = default;
+
+    Seq
+    start() const
+    {
+        return start_;
+    }
+
+    Seq
+    end() const
+    {
+        return end_;
+    }
+
+    // Return the Span from [spot,end_) or none if no such valid span
+    boost::optional<Span>
+    from(Seq spot) const
+    {
+        return sub(spot, end_);
+    }
+
+    // Return the Span from [start_,spot) or none if no such valid span
+    boost::optional<Span>
+    before(Seq spot) const
+    {
+        return sub(start_, spot);
+    }
+
+    // Return the ID of the ledger that starts this span
+    ID
+    startID() const
+    {
+        return ledger_[start_];
+    }
+
+    // Return the ledger sequence number of the first possible difference
+    // between this span and a given ledger.
+    Seq
+    diff(Ledger const& o) const
+    {
+        return clamp(mismatch(ledger_, o));
+    }
+
+    //  The Seq and ID of the end of the span
+    std::pair<Seq, ID>
+    tip() const
+    {
+        Seq tipSeq{end_ - Seq{1}};
+        return {tipSeq, ledger_[tipSeq]};
+    }
+
+private:
+    Span(Seq start, Seq end, Ledger const& l)
+        : start_{start}, end_{end}, ledger_{l}
+    {
+        // Spans cannot be empty
+        assert(start < end);
+    }
+
+    Seq
+    clamp(Seq val) const
+    {
+        return std::min(std::max(start_, val), end_);
+    };
+
+    // Return a span of this over the half-open interval [from,to)
+    boost::optional<Span>
+    sub(Seq from, Seq to) const
+    {
+        Seq newFrom = clamp(from);
+        Seq newTo = clamp(to);
+        if (newFrom < newTo)
+            return Span(newFrom, newTo, ledger_);
+        return boost::none;
+    }
+
+    friend std::ostream&
+    operator<<(std::ostream& o, Span const& s)
+    {
+        return o << s.tip().second << "[" << s.start_ << "," << s.end_ << ")";
```

Comment:
I actually prefer the look without the space. Do you strongly prefer with it?
```
0[0,1)(T:0,B:5)
                |-1[1,2)(T:0,B:5)
                                  |-2[2,3)(T:1,B:3)
                                                    |-6[3,5)(T:2,B:2)
                                  |-3[2,3)(T:1,B:2)
                                                    |-4[3,4)(T:1,B:1)
```

Reply:
Nope. As is is good.
What do you think of moving in the branch lines to the start of the parent's id?
```
0[0,1)(T:0,B:5)
|-1[1,2)(T:0,B:5)
  |-2[2,3)(T:1,B:3)
    |-6[3,5)(T:2,B:2)
  |-3[2,3)(T:1,B:2)
    |-4[3,4)(T:1,B:1)
```
Otherwise with actual ledger ids, this will get very wide very quickly.

Reply:
This is only  used by `dump`, which I only added to help build unit tests and its not actually called by anything currently.  For the ledger trie tests, I still find the wider version more readable.

If anything, you could convince me that deleting these functions is better since we aren't actually using them.

Reply:
in that case, it's fine as is